### PR TITLE
Add support for BlackWidow V3 (Full Size)

### LINKED ISSUE #1319: Support for BlackWidow V3 (Full Size)
### Device Info

* Name: Razer BlackWidow V3
* Product Number: RZ03-03540200-R3U1
* Model Number: RZ03-0354

### Device Information
```
Bus 001 Device 004: ID 1532:024e Razer USA, Ltd 
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x024e 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer BlackWidow V3
  iSerial                 3 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     179
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      79
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      22
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              2 Razer BlackWidow V3
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
Device Status:     0x0000
  (Bus Powered)
```
```
total 0
drwxr-xr-x 2 root root 240 Nov  2 18:02 .
drwxr-xr-x 4 root root 860 Nov  2 18:02 ..
lrwxrwxrwx 1 root root  10 Nov  2 18:02 usb-Razer_Razer_BlackWidow_V3-event-if01 -> ../event11
lrwxrwxrwx 1 root root  10 Nov  2 18:02 usb-Razer_Razer_BlackWidow_V3-event-if04 -> ../event15


lrwxrwxrwx 1 root root   9 Nov  2 18:02 usb-Razer_Razer_BlackWidow_V3-event-kbd -> ../event4
lrwxrwxrwx 1 root root   9 Nov  2 18:02 usb-Razer_Razer_BlackWidow_V3-if01-event-kbd -> ../event5
lrwxrwxrwx 1 root root  10 Nov  2 18:02 usb-Razer_Razer_BlackWidow_V3-if02-event-mouse -> ../event12
lrwxrwxrwx 1 root root   9 Nov  2 18:02 usb-Razer_Razer_BlackWidow_V3-if02-mouse -> ../mouse1
lrwxrwxrwx 1 root root   9 Nov  2 18:02 usb-Razer_Razer_DeathAdder_Elite-event-mouse -> ../event3
lrwxrwxrwx 1 root root  10 Nov  2 18:02 usb-Razer_Razer_DeathAdder_Elite-if01-event-kbd -> ../event23
lrwxrwxrwx 1 root root   9 Nov  2 18:02 usb-Razer_Razer_DeathAdder_Elite-if02-event-kbd -> ../event8
lrwxrwxrwx 1 root root   9 Nov  2 18:02 usb-Razer_Razer_DeathAdder_Elite-mouse -> ../mouse0
```
### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>
Properly archived.
[https://github.com/justZaduma/upload/blob/main/BWV3QuickOptionsEverySubOptionToggledInOrderTopToBottom.pcapng.tar.gz](BWV3QuickOptionsEverySubOptionToggledInOrderTopToBottom.pcapng.tar.gz)

### Synapse Screenshots
<Screenshots of all available options as seen on Razer Synapse>

image
image
image

### LINKED ISSUE #1593: Support for Razer BlackWidow V3
### Device Info

* Name: Razer BlackWidow V3
* Product Number: RZ03-03540400-R3G1
* Model Number: RZ03-0354

### Device Information
```
Bus 005 Device 003: ID 1532:024e Razer USA, Ltd Razer BlackWidow V3
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x024e 
  bcdDevice            2.00
  iManufacturer           1 Razer
  iProduct                2 Razer BlackWidow V3
  iSerial                 3 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0086
    bNumInterfaces          5
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     179
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0016  1x 22 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      79
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        3
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      0 
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      22
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x84  EP 4 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        4
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              2 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     348
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x85  EP 5 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0002  1x 2 bytes
        bInterval               1
```

```
lrwxrwxrwx 1 root root   9 Aug 26 11:32 usb-Razer_Razer_BlackWidow_V3-event-if01 -> ../event6
lrwxrwxrwx 1 root root   9 Aug 26 11:32 usb-Razer_Razer_BlackWidow_V3-event-if04 -> ../event8
lrwxrwxrwx 1 root root   9 Aug 26 11:32 usb-Razer_Razer_BlackWidow_V3-event-kbd -> ../event2
lrwxrwxrwx 1 root root   9 Aug 26 11:32 usb-Razer_Razer_BlackWidow_V3-if01-event-kbd -> ../event3
lrwxrwxrwx 1 root root   9 Aug 26 11:32 usb-Razer_Razer_BlackWidow_V3-if02-event-mouse -> ../event7
lrwxrwxrwx 1 root root   9 Aug 26 11:32 usb-Razer_Razer_BlackWidow_V3-if02-mouse -> ../mouse0
```

As far as i understand, it should be possible to base the implementation of this model on the existing Tenkeyless version.
If you do need additional information, please tell me and i'll set up a different rig to do reverse engineering.
Thanks in advance.

### PR DESCRIPTION
**Pull request by Spiffyk**
I have decided to try my hand at resolving #1319. The changes are mostly analogous to the Tenkeyless version of the keyboard (PR #1523), except for indices in `razer_get_report`, which I have successfully tested with value `0x03` (as opposed to `0x02` with the Tenkeyless).

All effects available in Polychromatic have been tested and seem to be working. I have also tested the custom matrix in RazerGenie, which mostly works, except for spacebar and right alt, but I have been told by Luca on Telegram that this is up for the GUIs to resolve.

**Known issue:** Key translations are incorrect:
 * `Fn+F11` is registered as microphone mute, when all it should do is reduce the keyboard backlight brightness
 * ~`Fn+F1`, `Fn+F2`, `Fn+F3` are registered as volume controls (mute, down, up; respectively), when the keyboard does not indicate it should do that (it has a dedicated volume scrollwheel)~ (fixed)
 * Media controls:
   * ~Scrolling the scrollwheel is registered as a mouse scrolling, when it should work as volume control~ (fixed)
   * ~Pushing the scrollwheel types "bcd", when it should mute audio~ (fixed)
   * ~Pushing the media button types "bcd", when it should act as `pause`, `next song`, `previous song` when pressed once, twice, or three times, respectively~ (fixed)
   * ~*These media controls worked correctly when openrazer was not in use, so there is probably some handling of this in the kernel and openrazer seems to override it?*~
 * *The issue gets fixed when the computer is put to sleep and restored. Rebooting or reconnecting the keyboard reverts it back however.* EDIT: caused by device mode vs. driver mode - see [comment](https://github.com/openrazer/openrazer/pull/1572#issuecomment-893598938)

Fixes #1319

### CODE REVIEW

**Comment by Spiffyk** [comment]
So I've done some more testing with the BlackWidow V3 media keys and I've found something interesting with their raw events:
  * when the computer is freshly booted, or the keyboard is physically reconnected, the first byte of the raw event has value `0x04`, which is what the openrazer driver matches against and produces the wrong behaviour
  * however, when I reload the razerkbd module (either via rmmod+modprobe, or sleep+wake cycle), the first byte of the raw event is `0x02`, which openrazer ignores and I assume this triggers some default behaviour in the kernel, which is actually correct for these buttons

**Comment by Spiffyk** [comment]
After further poking around and getting myself more acquainted with openrazer's architecture, I've now come to understand that the two states in which the keyboard sends events differently as described in my previous comments, are actually just the device mode and driver mode, which are already in one way or another handled by the driver and the daemon.

In device mode, the events are quite understandably correct, as (if I understand it correctly) the keyboard behaves as much according to the standard HID protocol as possible. So this part is fine.

Now in driver mode, we have some strange behaviour, because we are getting unwanted events:
  * Scrollwheel click should emit an event that the daemon interprets as `mute`. Instead, it emits three HID events as if `b`, `c` and `d` were typed.
  * Media button clicks should emit an event that the daemon interprets as `pause`, `next song`, or `previous song` according to the number of clicks. Instead, it again emits three HID events as if `b`, `c` and `d` were typed.
  * `Fn+F11` currently correctly emits an event that dims the keyboard's backlight, but it also emits an event interpreted by the daemon as `mic mute`.

So now I need to work out how to eliminate these events. I presume some more handling in `razer_raw_event` will be needed, but I'm not really sure about the format of the `data` array, so if there is anybody who could point me in the right direction (even just documentation links would be fine), please do hit me up.

Also, the daemon probably needs to have its probing adjusted, as it does not always attach itself to the keyboard (e.g. after the sleep/wake cycle, or when the daemon starts with the keyboard already connected).

**Comment by Spiffyk** [comment]
So I think I figured it out. Raw events starting with `0x01` that are 22 bytes in size are handled as bit fields, meaning that `razer_raw_event` was generating them incorrectly. I have corrected the issue and added additional handling for media keys, as those don't fit the 160-bit long bit field. So the media keys now seem to work correctly at all times. This should also fix media keys for #1552.

The `Fn+F11` muting the microphone seems to be a known issue (#579).

File: driver/razerkbd_driver.h [hunk 1]
```diff
@@ -77,6 +77,7 @@
 #define USB_DEVICE_ID_RAZER_CYNOSA_V2 0x025E
 #define USB_DEVICE_ID_RAZER_BOOK_2020 0x026A
 #define USB_DEVICE_ID_RAZER_BLADE_15_ADV_EARLY_2021 0x026D
+#define USB_DEVICE_ID_RAZER_BLACKWIDOW_V3 0x024E
```
Comment (z3ntu) [inline driver/razerkbd_driver.h]: Move this just below `USB_DEVICE_ID_RAZER_BLADE_STUDIO_EDITION_2019` to keep the list sorted

File: driver/usb_hid_keys.h [hunk 2]
```diff
@@ -25,4 +25,11 @@
 #define USB_HID_KEY_MEDIA_MUTE 0xEF
 #define USB_HID_KEY_MEDIA_SLEEP 0xF8
 
+#define USB_HID_USAGE_MEDIA_PLAYPAUSE 0xCD
+#define USB_HID_USAGE_MEDIA_PREVIOUSSONG 0xB6
+#define USB_HID_USAGE_MEDIA_NEXTSONG 0xB5
+#define USB_HID_USAGE_MEDIA_VOLUMEUP 0xE9
+#define USB_HID_USAGE_MEDIA_VOLUMEDOWN 0xEA
+#define USB_HID_USAGE_MEDIA_MUTE 0xE2
```
Comment (z3ntu) [inline driver/usb_hid_keys.h:34]: Can you please add a reference where these numbers are from?

File: driver/razerkbd_driver.c [hunk 3]
```diff
@@ -2317,15 +2356,68 @@ static int razer_raw_event(struct hid_device *hdev, struct hid_report *report, u
...
+
+                            // report key up
+                            xdata[1] = 0x00;
+                            hid_report_raw_event(hdev, HID_INPUT_REPORT, xdata, sizeof(xdata), 0);
+                        }
+                    }
+                }
+            } else { // size 16
+                data[index+1] = cur_value;
+            }
```
Comment (z3ntu) [inline driver/razerkbd_driver.c:2745]: I'm definitely not happy with modifying this function as I have neither a reference for it how it should work and we had too many regressions with this in the past...

File: driver/razerkbd_driver.h (hunk 1 above)
Comment (Spiffyk) [reply driver/razerkbd_driver.h]: Moved.

File: driver/usb_hid_keys.h (hunk 2 above)
Comment (Spiffyk) [reply driver/usb_hid_keys.h:34]: Added reference to official USB HID usage document: https://usb.org/sites/default/files/hut1_22.pdf

File: driver/razerkbd_driver.c (hunk 3 above)
Comment (Spiffyk) [reply driver/razerkbd_driver.c:2745]: Yes, that's understandable. I wasn't able to find any good documentation for this either, and I only figured the behaviour out by trial end error. It is evident however, that this change would fix multiple keyboards, as Ornata Chroma V2 has the same incorrect behaviour as BlackWidow V3 before the change (see issue #1552).

Maybe the correct approach would be to add another switch statement that differentiates by ProductID here. What do you think?

**Comment by Spiffyk** [comment]
Rebased onto current `master`. Cannot do a sanity check right now, as I'm out of town where I have my keyboard, sadly.

**Comment by prosaied** [comment]
Spiffyk, thank you so much for your all efforts, is there any chance that we could test your version?

**Comment by Spiffyk** [comment]
@prosaied Hi, you can build it on your own if you want. Just `git clone` my forked repo, `git checkout origin/feature_blackwidowv3fullsize`, then use [this page as a reference to build and install your package from sources](https://github.com/openrazer/openrazer/wiki/Building-a-package).

**Comment by KitsuneAlex** [comment]
@Spiffyk thanks a ton, was about to implement this on my own 😆 Much love <3

File: driver/razerkbd_driver.c (hunk 3 above)
Comment (z3ntu) [reply driver/razerkbd_driver.c:2745]: That's the sanest solution I can think of for now. It probably makes sense to split the code for this device out into its own function and call it from here, so the function doesn't get a million lines long.

**Comment by prosaied** [comment]
@Spiffyk I have installed with your forked repo, and it is working like a charm on my Ubuntu 20.04 installation.
Thank you so much.

**Comment by federicogon** [comment]
It's working great to me too. Ubuntu 20.04.3 LTS

**Comment by TheDiscordian** [comment]
Manjaro 21.1.1, no issues to report. Thank you for your work! :pray:

**Comment by geodox** [comment]
Unfortunately, it's not working for my full size on Pop OS 21.04
image

**Comment by TheDiscordian** [comment]
@GeoDoX troubleshooter provide any hints? I had to re-add myself to a group, load a kernel module, and reboot. After doing all that and then telling Polychromatic to reload the daemon, it worked like a charm.

**Comment by geodox** [comment]
This was the output from the troubleshooter. I have restarted the system, and I only have this branches OpenRazer installed. No issues building the debs. I already have the plugdev group on my user. Can you provide more details about loading the kern module? sudo modprobe razerkbd outputs nothing.
image

**Comment by TheDiscordian** [comment]
@GeoDoX the instructions are there, simply run `sudo modprobe razerkbd` (as you've done, no output just means there was no error), then restart the daemon via Polychromatic, then I believe you should be good to go. After that the keyboard should show up, but if it doesn't, the troubleshooter should *not* show "Failed" anymore (confirming the module is indeed loaded at least).

**Comment by geodox** [comment]
@TheDiscordian Still nothing. Since restarting the daemon, the dkms module passed and now it only fails on device compatibility.

**Comment by TheDiscordian** [comment]
@GeoDoX I just noticed your PID:VID is different from mine too, idk if that's significant. I believe mine is `1532:024E`, are you sure this PR is for your keyboard? My PID:VID is the same as listed in #1319, I wonder if your model is slightly different.

**Comment by Spiffyk** [comment]
@GeoDoX Your keyboard's PID seems to be `1532:025A`, that's not the fullsize BlackWidow V3 I have, mine has `1532:024E` and the PR won't work with any other PID. Are you sure you don't have one of the other editions of the keyboard?

**Comment by Spiffyk** [comment]
@GeoDoX Found it, it's the Pro, you need to look at #1514, this one's for the non-Pro full size.

**Comment by GeoDoX** [comment]
@Spiffyk `sudo lsusb` provides the following output.
`Bus 012 Device 002: ID 1532:025a Razer USA, Ltd Razer BlackWidow V3 Pro`

**Comment by Spiffyk** [comment]
@z3ntu I've added the branching of `razer_raw_event` by `productId`. Do you want me to also add the case for Ornata Chroma V2, or do you wish to handle that separately with someone who has that keyboard?

**Comment by KamranMackey** [comment]
@Spiffyk regarding the volume controls being bound to function keys, that's actually proper behaviour - while the default keycaps that come with the Blackwidow V3 dont indicate it, when you have Synapse running, Fn keys F1 through F3 control volume functions. If you install the PBT keycaps, they show the proper visuals.

**Comment by krpanko** [comment]
woow guys, this looks great. Thanks. Looking forward to merge and testing the release. 

Thanks and big KUDOS to all here. OSS for life :muscle:

**Comment by krpanko** [comment]
woow guys, this looks great. Thanks. Looking forward to merge and testing the release. 

Thanks and big KUDOS to all here. OSS for life :muscle:  

wen :)

**Review by krpanko — APPROVED** [review]

**Comment by logical-luke** [comment]
In general it seems to be working great, good job :muscle: 
I found one problem that at least for me is not a big deal as I'm not using **Gaming Mode** at all.
When switching **Gaming Mode** in Polychromatic it throws error, but it seems to be toggling it in both ways. 
When switching it using `FN+F10` we can only enable this mode, but I wasn't able to disable it.

image
			
```
Traceback (most recent call last):
  File ’/usr/lib/python3/dist-packages/polychromatic/backends/openrazer.py’, line 978, in set_device_state
    rdevice.game_mode_led = option_data
  File ’/usr/lib/python3/dist-packages/openrazer/client/devices/keyboard.py’, line 53, in game_mode_led
    self._dbus_interfaces[’game_mode_led’].setGameMode(True)
  File ’/usr/lib/python3/dist-packages/dbus/proxies.py’, line 141, in __call__
    return self._connection.call_blocking(self._named_service,
  File ’/usr/lib/python3/dist-packages/dbus/connection.py’, line 652, in call_blocking
    reply_message = self.send_message_with_reply_and_block(
dbus.exceptions.DBusException: org.freedesktop.DBus.Python.PermissionError: Traceback (most recent call last):
  File ’/usr/lib/python3/dist-packages/dbus/service.py’, line 711, in _message_cb
    retval = candidate_method(self, *args, **keywords)
  File ’/usr/lib/python3/dist-packages/openrazer_daemon/dbus_services/dbus_methods/chroma_keyboard.py’, line 85, in set_game_mode
    open(super_file, ’wb’).write(b’\x01’)
PermissionError: [Errno 13] Permission denied: ’/sys/devices/pci0000:00/0000:00:14.0/usb1/1-8/1-8:1.3/0003:1532:024E.0005/key_super’
```

**Comment by iparaskev** [comment]
@Spiffyk I added the Ornata V2 product id in your patch and tested it on my keyboard. It is safe to include it as everything works fine.

**Comment by devwin03** [comment]
Hey there - so, I'm not all that experienced with these sorts of things, and have been trying to figure out how to install this for about three hours now. These instructions [https://github.com/openrazer/openrazer/wiki/Building-a-package] are vague, and a bit of guidance about what exactly to edit into the openrazer.spec file would be a godsend (I'm on fedora, if that helps)

**Comment by KitsuneAlex** [comment]
> Hey there - so, I'm not all that experienced with these sorts of things, and have been trying to figure out how to install this for about three hours now. These instructions [https://github.com/openrazer/openrazer/wiki/Building-a-package] are vague, and a bit of guidance about what exactly to edit into the openrazer.spec file would be a godsend (I'm on fedora, if that helps)

Hmu on Discord if you need some help. KitsuneAlex#0001

**Comment by Spiffyk** [comment]
@z3ntu Just to make sure I'm not unknowingly blocking this - is there anything more I should do (apart from a rebase on `master`) to get this PR merged?

**Comment by Acidpunk** [comment]
Does anyone have any advice on how to build the forked driver using an Arch based distro, I couldn't get makpkg to run successfully yesterday trying.

**Comment by z3ntu** [comment]
@Spiffyk Can you update me on the list of known issues from the first comment? 

Fn+F11 = mic-mute is a generally known so you can ignore that. But what about Fn+F1, Fn+F2, Fn+F3 ?

**Comment by Spiffyk** [comment]
@z3ntu Sorry, totally forgot to update the issue description! `Fn+F1`, `Fn+F2`, and `Fn+F3` are fixed by the separate `chroma_keys_5` table.

**Comment by Spiffyk** [comment]
@Acidpunk I use Manjaro and it works. What message does `makepkg` fail with? I initially had a problem that I added the `prepare()` function without realizing that newer versions of the `PKGBUILD` file already have one, so you need to add the `cp` command to that.

**Comment by Acidpunk** [comment]
> @Acidpunk I use Manjaro and it works. What message does `makepkg` fail with? I initially had a problem that I added the `prepare()` function without realizing that newer versions of the `PKGBUILD` file already have one, so you need to add the `cp` command to that.

Yeah I was getting issues with prepare() and later with something python related I'll give it another crack tomorrow then and try what you said

**Comment by michaellee8** [comment]
@Spiffyk When will this be released? Looks very useful for me. I just need to disable all lights on my keyboard.

**Comment by phl0** [comment]
Would this patch also work for 1532:025A if I added the PID? My blackwidow v3 pro says it has these IDs ...

**Comment by Spiffyk** [comment]
@michaellee8 I'm not the right person to ask, it's not up to me

@phl0 There are separate pull requests for the BlackWidow V3 Pro: #1622 #1623

**Comment by geodox** [comment]
@phl0 please test #1622, comment on that PR with good or bad feedback please :)

**Comment by phl0** [comment]
@GeoDoX @Spiffyk Thanks info. Tested with positive result.  #1622 is good to go :)

https://github.com/openrazer/openrazer/pull/1622#issuecomment-947534243

**Comment by gediz** [comment]
Thank you so much, @Spiffyk

I have tested it on Ubuntu 20.04.3 LTS and it works seamlessly. For those who want to try it, you can run the following script

```bash
#!/bin/bash

# clone the repo and cd to it
git clone -b feature_blackwidowv3fullsize https://github.com/Spiffyk/openrazer.git
cd openrazer || exit 1

# install dependencies
sudo apt install debhelper dh-python linux-headers-generic python3-setuptools lsb-release dkms

# build it
./scripts/build_debs.sh

# install compiled deb packages
sudo apt install ./dist/*.deb

# add user to the plugdev group
sudo gpasswd -a "$USER" plugdev
```

I had to restart my PC but maybe it'd work without a restart by removing and probing the module again, idk.

openrazer_razergenie_blackwidowv3fullsize

**Review by z3ntu — APPROVED** [review]
LGTM, thanks for the pull request! And thanks for testing, everybody!

**Comment by jake-arkinstall** [comment]
Thanks for this PR and all the subsequent work on getting it finalised. I'm looking forward to trying it out!

**Pull request merged**